x/stake/cross_stake/cross_stake.go

```diff
@@ -0,0 +1,340 @@
+	sideChainId := app.stakeKeeper.DestChainName
+	if scCtx, err := app.stakeKeeper.ScKeeper.PrepareCtxForSideChain(ctx, sideChainId); err != nil {
+		return sdk.ExecuteResult{}, err
+	} else {
+		ctx = scCtx
+	}
+
+	valDst, found := app.stakeKeeper.GetValidator(ctx, pack.ValDst)
+	if !found || valDst.Jailed {
+		var sdkErr sdk.Error
+		var errCode uint8
+		if !found {
+			sdkErr = types.ErrNoValidatorFound(types.DefaultCodespace)
+			errCode = CrossStakeErrValidatorNotFound
+		} else {
+			sdkErr = types.ErrValidatorJailed(types.DefaultCodespace)
+			errCode = CrossStakeErrValidatorJailed
+		}
+		ackPack := types.NewCrossStakeRedelegationAckPackage(&pack, errCode)
+		ackPack.Amount = bsc.ConvertBCAmountToBSCAmount(ackPack.Amount.Int64())
+		ackBytes, err := rlp.EncodeToBytes(ackPack)
+		if err != nil {
+			return sdk.ExecuteResult{}, err
+		}
+		return sdk.ExecuteResult{
+			Err:     sdkErr,
+			Payload: ackBytes,
+		}, nil
+	}
+
+	delAddr := types.GetStakeCAoB(pack.DelAddr[:], "Delegate")
+	shares, sdkErr := app.stakeKeeper.ValidateUnbondAmount(ctx, delAddr, pack.ValSrc, pack.Amount.Int64())
+	if sdkErr != nil {
+		ackPack := types.NewCrossStakeRedelegationAckPackage(&pack, CrossStakeErrBadDelegation)
+		ackPack.Amount = bsc.ConvertBCAmountToBSCAmount(ackPack.Amount.Int64())
+		ackBytes, err := rlp.EncodeToBytes(ackPack)
+		if err != nil {
+			return sdk.ExecuteResult{}, err
+		}
+		return sdk.ExecuteResult{
+			Err:     sdkErr,
+			Payload: ackBytes,
+		}, nil
+	}
+
+	_, err := app.stakeKeeper.BeginRedelegation(ctx.WithCrossStake(true), delAddr, pack.ValSrc, pack.ValDst, shares)
+	if err != nil {
+		return sdk.ExecuteResult{}, err
+	}
+
+	// publish redelegate event
+	if app.stakeKeeper.PbsbServer != nil && ctx.IsDeliverTx() {
+		event := types.SideRedelegateEvent{
+			RedelegateEvent: types.RedelegateEvent{
+				StakeEvent: types.StakeEvent{
+					IsFromTx: true,
+				},
+				Delegator:    delAddr,
+				SrcValidator: pack.ValSrc,
+				DstValidator: pack.ValDst,
+				Amount:       pack.Amount.Int64(),
+				Denom:        app.stakeKeeper.BondDenom(ctx),
+				TxHash:       ctx.Value(baseapp.TxHashKey).(string),
+			},
+			SideChainId: sideChainId,
+		}
+		app.stakeKeeper.PbsbServer.Publish(event)
+		PublishCrossChainEvent(ctx, app.stakeKeeper, delAddr, pack.ValSrc, pack.ValDst, types.CrossStakeRedelegateType,
+			relayFee)
+	}
+
+	resultTags := sdk.NewTags(
+		types.TagCrossStakePackageType, []byte(strconv.FormatInt(int64(types.CrossStakeTypeRedelegate), 10)),
+	)
+	return sdk.ExecuteResult{
+		Tags: resultTags,
+	}, nil
+}
+
+func (app *CrossStakeApp) handleDistributeRewardRefund(ctx sdk.Context, pack types.CrossStakeDistributeRewardSynPackage) (sdk.ExecuteResult, error) {
+	symbol := app.stakeKeeper.BondDenom(ctx)
+	refundAmount := bsc.ConvertBSCAmountToBCAmount(pack.Amount)
+	coins := sdk.Coins{sdk.NewCoin(symbol, refundAmount)}
+	delAddr := types.GetStakeCAoB(pack.Recipient[:], "Delegate")
```

Comment:
Please either mark  "Delegate" and  "Reward" as constant, or implement `GetStakeDelCAoB` and `GetStakeRewardCAoB`.

Reply:
Done